github.com/cockroachdb/pebble/internal/metamorphic: TestMeta failed

github.com/cockroachdb/pebble/internal/metamorphic.TestMeta [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_Pebble_Metamorphic/17234170?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_Pebble_Metamorphic/17234170?buildTab=artifacts#meta) on refs/heads/master @ [e1b330bd4290](https://github.com/cockroachdb/pebble/commits/e1b330bd4290):


```
        iter32.Prev("") // [true,"jvmuweuv@13","xvnkooi",<no range>] <nil> #2714
        iter32.Prev("") // [true,"jvmuweuv@15","gdx",<no range>] <nil> #2716
        iter32.SetBounds("", "djgkk") // <nil> #2720
        iter32.Prev("") // [false] <nil> #2728
        iter32.Last() // [true,"ddjmtk@11","dpsxbru",["crjyzgxyfz","djgkk")=>{"@25"="khcjmz"}*] <nil> #2729
        iter32.Last() // [true,"ddjmtk@11","dpsxbru",["crjyzgxyfz","djgkk")=>{"@25"="khcjmz"}] <nil> #2735
        iter32.Prev("") // [true,"ddjmtk@17","yqjsljrjeirfvsrqzy",["crjyzgxyfz","djgkk")=>{"@25"="khcjmz"}] <nil> #2737
        db1.Get("jwmltti@8") // [""] pebble: not found #2389
        db1.LogData("mqn") // <nil> #2391
        db1.LogData("") // <nil> #2405
        batch27 = db1.NewIndexedBatch() #2411
        db1.RangeKeyUnset("crjyzgxyfz", "jwmltti", "@11") // <nil> #2414
        batch27.Close() // <nil> #2417
        db1.Get("azuf@24") // ["luyvmsdmaedvdjsgsbzh"] <nil> #2429
        db1.Set("hxjbcwdmqc@1", "bpmeglcpdiqhtf") // <nil> #2430
        db1.Get("ucok@25") // ["lwxn"] <nil> #2434
        db1.Set("cezrfazxxkz@27", "lahohjeoqit") // <nil> #2437
        db1.Set("gmxgelriepfk", "lxzwoxjxowloqt") // <nil> #2438
        db1.RangeKeyUnset("ljdjb", "ucok", "@27") // <nil> #2439
        db1.Get("jwmltti@21") // ["giuh"] <nil> #2441
        db1.Set("jwmltti@10", "") // <nil> #2443
        db1.Get("pxvs") // ["gczu"] <nil> #2444
        db1.Set("pxvs@21", "lowpewwvuikwxi") // <nil> #2454
        db1.Set("kaeue@27", "spucpsntb") // <nil> #2461
        db1.Set("gmxgelriepfk@1", "") // <nil> #2466
        db1.Get("ljdjb@2") // ["orlsfkduynp"] <nil> #2469
        db1.Get("iusttwznud@24") // ["nsiryaabi"] <nil> #2475
        db1.Set("ddjmtk@16", "gw") // <nil> #2477
        db1.Set("cfkb@15", "jlpoux") // <nil> #2478
        db1.Set("djgkk@18", "lwakuzjebwzhw") // <nil> #2496
        db1.Get("iusttwznud@20") // ["jdp"] <nil> #2500
        db1.Get("gmxgelriepfk@14") // [""] pebble: not found #2505
        db1.RangeKeySet("crjyzgxyfz", "kpnkjyy", "@25", "mqvoxkjumcvfhmj") // <nil> #2514
        db1.RangeKeySet("evsftdsvz", "yhbnazwnvkfo", "@26", "kqrqqzz") // <nil> #2515
        db1.Set("iusttwznud@26", "erywdmiztdtodlm") // <nil> #2518
        db1.Get("omitbnecpnjy@2") // ["quyihtdaljmuu"] <nil> #2521
        db1.Set("kpatwqczyzjw@24", "anshgbnyg") // <nil> #2529
        snap26 = db1.NewSnapshot("gdntxggueed", "uxnopz") #2530
        db1.Set("jvmuweuv@22", "ywixfumvdderrblb") // <nil> #2532
        db1.Get("zjswazj") // [""] pebble: not found #2548
        db1.LogData("epqg") // <nil> #2552
        db1.Set("uqkejjtzx", "") // <nil> #2557
        snap27 = db1.NewSnapshot("bwczvs", "cezrfazxxkz", "cfkb", "djgkk", "evsftdsvz", "hpxn", "hxjbcwdmqc", "jtagaccewmfn", "ljdjb", "txeeahmkl", "ucok", "uezln", "ufdhwimk", "uxnopz") #2558
        db1.Set("pmenq@24", "oqkuzpuqqfqd") // <nil> #2560
        db1.Set("azuf@24", "ue") // <nil> #2561
        
        To reduce:  go test ./internal/metamorphic -tags invariants -run 'TestMeta$' --run-dir /artifacts/meta/241009-042933.8602830867539/standard-013 --try-to-reduce -v
        --- FAIL: TestMeta/execution/standard-013 (0.30s)
=== RUN   TestMeta/execution
    --- FAIL: TestMeta/execution (0.00s)
```
<details><summary>Help</summary>
<p>
To reproduce, try:

```bash
go test -tags 'invariants' -exec 'stress -p 1' -timeout 0 -test.v -run 'TestMeta$' ./internal/metamorphic -seed 1728448173860773011 -ops "uniform:5000-10000"
```
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMeta.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>


Jira issue: PEBBLE-272

github.com/cockroachdb/pebble/internal/metamorphic.TestMeta [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicRace/17234169?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_NightlyPebbleMetamorphicRace/17234169?buildTab=artifacts#meta) on refs/heads/master @ [e1b330bd4290](https://github.com/cockroachdb/pebble/commits/e1b330bd4290):


```
        iter23.Next("") // [true,"rwpxybgvrg",<no point>,["rwpxybgvrg","trenpzjmti")=>{""="tfohcvqtoktpwjvyz"}*] <nil> #1930
        iter23.SeekPrefixGE("varyqabh@19") // [true,"varyqabh@19","gerqg",<no range>*] <nil> #1934
        iter23.Next("") // [false] <nil> #1949
        iter23.SetOptions("irjzolvueqn@6", "rwpxybgvrg", 2 /* key types */, 0, 0, false /* use L6 filters */, "" /* masking suffix */) // <nil> #1953
        iter23.SeekLT("rmkhi", "qpyjovt@13") // [valid,"rlhptkohyg",<no point>,["rlhptkohyg","rmkhi")=>{""="tfohcvqtoktpwjvyz","@19"="lefzdi"}*] <nil> #1954
        iter23.SeekLT("jjft", "exwcfc@15") // [valid,"izfkmlmg@9","vlxrstckznfyykwwfe",<no range>*] <nil> #1961
        iter23.SeekGE("lkjw@13", "trenpzjmti@17") // [valid,"lkjw@13","hqgelfxbznghd",<no range>] <nil> #1964
        iter23.First() // [true,"izfkmlmg@9","vlxrstckznfyykwwfe",<no range>] <nil> #1967
        iter23.SetBounds("exwcfc@15", "fhpooezterz@11") // <nil> #1973
        iter23.SeekGE("exwcfc@15", "") // [false] <nil> #1974
        iter23.Prev("") // [false] <nil> #1975
        iter23.Prev("") // [false] <nil> #1989
        iter23.NextPrefix() // [false] NextPrefix not permitted with upper bound fhpooezterz@11 #1995
        iter23.SeekPrefixGE("trenpzjmti@17") // [false] pebble: SeekPrefixGE supplied with key outside of upper bound #1997
        iter23.Prev("rwpxybgvrg@20") // [invalid] pebble: SeekPrefixGE supplied with key outside of upper bound #2004
        iter23.Next("") // [false] pebble: SeekPrefixGE supplied with key outside of upper bound #2008
        iter18.Last() // [true,"banhyvhox","qywklvplalnsasutm",["banhyvhox","banhyvhox@10")=>{"@6"="gamwoaxojbdwqgjzo"}] <nil> #2110
        iter18.Last() // [true,"banhyvhox","qywklvplalnsasutm",["banhyvhox","banhyvhox@10")=>{"@6"="gamwoaxojbdwqgjzo"}] <nil> #2112
        db1.Get("rlhptkohyg@14") // [""] pebble: not found #1694
        db1.Set("yvdmz@12", "rocuimhan") // <nil> #1706
        db1.Get("ljyretuzp@8") // [""] pebble: not found #1708
        iter18.First() // [true,"banhyvhox","qywklvplalnsasutm",["banhyvhox","banhyvhox@10")=>{"@6"="gamwoaxojbdwqgjzo"}] <nil> #2120
        iter18.Next("") // [false] <nil> #2121
        db1.Set("idsck@12", "eoqedb") // <nil> #1709
        db1.Get("vvvsy@14") // [""] pebble: not found #1710
        iter18.First() // [true,"banhyvhox","qywklvplalnsasutm",["banhyvhox","banhyvhox@10")=>{"@6"="gamwoaxojbdwqgjzo"}*] <nil> #2128
        db1.Set("pjez@8", "ceyavbfdmhsxldcxiev") // <nil> #1711
        db1.Set("exwcfc@19", "zliyczfkeatsowr") // <nil> #1714
        iter18.Last() // [true,"banhyvhox","qywklvplalnsasutm",["banhyvhox","banhyvhox@10")=>{"@6"="gamwoaxojbdwqgjzo"}] <nil> #2131
        iter18.SetBounds("", "rwpxybgvrg@20") // <nil> #2142
        db1.Get("qufvyo@5") // [""] pebble: not found #1716
        db1.Get("vvvsy@3") // [""] pebble: not found #1718
        iter11.SeekLT("zuabtzeppsod@19", "rlhptkohyg@14") // [invalid] <nil> #1733
        iter23.First() // [false] <nil> #2013
        iter18.Next("") // [true,"banhyvhox","qywklvplalnsasutm",["banhyvhox","dvucftywqau")=>{"@6"="gamwoaxojbdwqgjzo"}*] <nil> #2159
        iter18.SetBounds("", "jyeaknxagx@17") // <nil> #2171
        iter23.SeekPrefixGE("vvvsy@5") // [false] pebble: SeekPrefixGE supplied with key outside of upper bound #2026
        db1.Get("jyeaknxagx") // [""] pebble: not found #1720
        db1.Get("pbxmrhv@17") // [""] pebble: not found #1725
        iter19.SeekGE("kzqqr", "") // [false] <nil> #1742
        iter18.SeekLT("ydlnvqxswh@16", "trenpzjmti@17") // [invalid] <nil> #2177
        iter18.SetBounds("njfqdipbzpao@17", "pjez@7") // <nil> #2181
        iter18.Prev("") // [false] <nil> #2183
        iter21 = db1.NewIter("", "", 1 /* key types */, 0, 0, false /* use L6 filters */, "" /* masking suffix */) // <nil> #1738
        iter18.Last() // [true,"pjez@14","hv",<no range>] <nil> #2187
        iter23.Prev("dylr@15") // [invalid] pebble: SeekPrefixGE supplied with key outside of upper bound #2039
        iter19.First() // [true,"fhpooezterz@5",<no point>,["fhpooezterz@5","irjzolvueqn")=>{""="maxvtt","@6"="gamwoaxojbdwqgjzo"}*] <nil> #1743
        
        To reduce:  go test ./internal/metamorphic -tags invariants -run 'TestMeta$' --run-dir /artifacts/meta/241009-044846.250104504805/random-010 --try-to-reduce -v
        --- FAIL: TestMeta/execution/random-010 (0.80s)
```
<details><summary>Help</summary>
<p>
To reproduce, try:

```bash
go test -tags 'invariants' -exec 'stress -p 1' -timeout 0 -test.v -run 'TestMeta$' ./internal/metamorphic -seed 1728449326250683419 -ops "uniform:5000-10000"
```
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMeta.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)
</sub>

--- a/iterator.go
+++ b/iterator.go
@@ -817,8 +817,10 @@ func (i *Iterator) maybeSampleRead() {
 		// every newly-opened iterator, but we do want to sample some of them.
 		if !i.readSampling.initialSamplePassed {
 			i.readSampling.initialSamplePassed = true
-			if rand.Uint32N(uint32(i.readSampling.bytesUntilReadSampling)) > uint32(bytesRead) {
-				continue
+			if i.readSampling.bytesUntilReadSampling > bytesRead {
+				if rand.Uint64N(i.readSampling.bytesUntilReadSampling) > bytesRead {
+					continue
+				}
 			}
 		}
 		i.sampleRead()

